### PR TITLE
Avoid using skip() in hf_datasets

### PR DESCRIPTION
**Pull request by mori360**
Fix issue https://github.com/pytorch/torchtitan/issues/809

The current `self._data.skip(self._sample_idx)` could not get the correct data for c_4 dataset.
Thus we switch to next() first before the fix is landed.

Test plan:
We reproduce the https://github.com/pytorch/torchtitan/issues/809 by resuming from checkpoint at step 500, then compare the loss curve in 3 conditions:
1. the origin curve running from step 0 to 750
2. the resumed curve keeping .skip()
3. the resumed curve switch to next(), with this PR change
Screenshot 2025-02-12 at 11 19 24 AM

Warning
for c_4 dataset, if we resume from a large enough step, we call next() for `self._sample_idx` times, resuming from checkpoint would be much slower than using .skip()

Next step:
add unit test:
1. test the state_dict check between dcp.save/load and torch.save/load
2. test the difference between next() and .skip()

### CODE REVIEW

**Comment by mori360** [comment]
Screenshot 2025-02-12 at 11 48 36 AM
I have `pre-commit run --files torchtitan/datasets/hf_datasets.py` locally and they all pass.

File: torchtitan/datasets/hf_datasets.py [hunk 1]
```diff
@@ -99,13 +99,13 @@ def __init__(
         self._all_tokens: List[int] = []
 
     def _get_data_iter(self):
-        if self._sample_idx == 0:
-            return iter(self._data)
-
         if isinstance(self._data, Dataset) and self._sample_idx == len(self._data):
             return iter([])
 
-        return iter(self._data.skip(self._sample_idx))
```
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py:108]: I think we need to understand if `skip` causes error in both map-style and Iterable datasets, or only in the newly added IterableDataset case.
If it's the latter we should just revert https://github.com/pytorch/torchtitan/pull/521, rather than universally use `next` for both, because it would make the healthy case slow too.

Comment (fegin) [reply torchtitan/datasets/hf_datasets.py:108]: I would suggest that we land the PR first. It is better to have a slower checkpoint resume than an incorrect silent accuracy failure. It's blocking several accuracy verifications. Or at least we should make the default C4 dataset work for now.

**Review by tianyu-l — APPROVED** [review]
stamp to unblock, but we should follow up with more robust tests.
